wp/lib/bap_wp/tests/unit/test_cfg_path.ml

```diff
@@ -0,0 +1,310 @@
+open !Core
+open Bap.Std
+open OUnit2
+open Testing_utilities
+open Bap_wp
+
+module Pre = Precondition
+module Comp = Compare
+module Constr = Constraint
+module Env = Environment
+module Bool = Z3.Boolean
+module BV = Z3.BitVector
+module Out = Output
+module CP = Cfg_path
+module Node = Graphs.Ir.Node
+module Edge = Graphs.Ir.Edge
+module EdgeSet = Edge.Set
+module Graph = Graphs.Ir
+
+let code_of_sub sub env = Comp.{
+    env;
+    prog = sub;
+    mem = Memmap.empty;
+    code_addrs = Utils.Code_addrs.empty; }
+
+let string_of_edgeset (es : EdgeSet.t) : string =
+  EdgeSet.fold es ~init:"\n" ~f:(fun s e -> s ^ Graphs.Ir.Edge.to_string e ^ "\n")
+
+let edge_of_endpoints (src : Blk.t) (dst : Blk.t) : Edge.t =
+  let cfg = [src; dst] |> mk_sub |> Sub.to_cfg in
+  let es = Graph.edges cfg |> Seq.to_list in
+  match es with
+  | [e] -> e
+  | _   -> failwith "CFG should consist of a single edge"
+
+(* TODO : refactor tests below in terms of this function *)
+let validate_taken_edges (sub1 : Sub.t) (sub2 : Sub.t)
+      ~(pre_regs : Var.Set.t) ~(post_regs : Var.Set.t)
+    ~(expected1 : EdgeSet.t) ~(expected2 : EdgeSet.t) : unit =
+  let ctx = Env.mk_ctx () in
+  let var_gen = Env.mk_var_gen () in
+  let env1 = Pre.mk_env ~target:test_tgt ctx var_gen in
+  let env2 = Pre.mk_env ~target:test_tgt ctx var_gen in
+  let post, hyps = Comp.compare_subs_eq ~pre_regs ~post_regs in
+  let compare_prop, env1, env2 = Comp.compare_subs
+      ~postconds:[post] ~hyps:[hyps]
+      ~original:(code_of_sub sub1 env1) ~modified:(code_of_sub sub2 env2) in
+  let z3_ctx = Env.get_context env1 in
+  let solver = Z3.Solver.mk_simple_solver z3_ctx in
+  let result = Pre.check solver z3_ctx compare_prop in
+  match result with
+  | Solver.UNSATISFIABLE -> failwith "err"
+  | Solver.UNKNOWN -> failwith "unk"
+  | Solver.SATISFIABLE ->
+     let mem1, _ = Env.get_var env1 (Env.get_mem env1) in
+     let mem2, _ = Env.get_var env2 (Env.get_mem env2) in
+     let refuted_goals =
+       Constr.get_refuted_goals compare_prop solver z3_ctx ~filter_out:[mem1; mem2] in
+     match Seq.hd refuted_goals with
+     | None -> assert_failure "expected at least one refuted goal after SAT result"
+     | Some rg ->
+        let es1, es2 = CP.taken_edges_of_refuted_goal rg sub1 sub2 in
+        let assert_equal_sets (exp : EdgeSet.t) (act : EdgeSet.t) =
+          assert_equal exp act ~cmp:EdgeSet.equal ~printer:string_of_edgeset in
+        begin
+          assert_equal_sets expected1 es1;
+          assert_equal_sets expected2 es2
+        end
+       
+(** 
```

Comment:
Consider putting human-readable versions of subroutines side-by-side instead of top-to-bottom.

Reply:
Done.